crontab.1
.\" Copyright 1994 Matthew Dillon (dillon@apollo.backplane.com)
.\" May be distributed under the GNU General Public License
.TH CRONTAB 1 "16 August 2026" "dcron 3.3-dev"
.SH NAME
crontab \- manipulate per-user crontabs (Dillon's Cron)
.SH SYNOPSIS
.B crontab
.RI [ SUPERUSER-OPTION ]...
.RB [ -e | -l | -r ]
.br
.B crontab
.RI [ SUPERUSER-OPTION ]...
.RI [ FILE ]
.PP
.B SUPERUSER-OPTIONS:
.TP
.BR -u \ user
specify crontab user
.TP
.BR -c \ dir
specify crontab directory
.SH DESCRIPTION
.B crontab
manipulates the crontab for a particular user.  Only the superuser may 
specify a different user and/or crontab directory.
.PP
The \fB\-e\fR option is used to edit a crontab and will use
/usr/bin/vi or the editor specified by the VISUAL environment
variable to edit the crontab.
.PP
The \fB\-l\fR option is used to list the contents of a crontab.
.PP
The \fB\-r\fR option is used to remove a crontab.
.PP
In the second form, where non of the options \fB\-e\fR, \fB\-l\fR or
\fB\-r\fR are specified, the contents of \fIFILE\fR will be used to
replace a crontab. If FILE is not specified or specified as
.RB ` \- '
input will be read from stdin.
.PP
When a crontab command is run, this cron implementation runs it with
/bin/sh and an environment comprising of only four environment
variables: HOME, LOGNAME, SHELL and PATH.
.PP
crond automatically detects changes in the time.  Reverse-indexed
time changes less then an hour old will NOT re-run crontab commands
already issued in the recovered period. Forward-indexed changes less 
then an hour
into the future will issue missed commands exactly once.  Changes
greater then an hour into the past or future cause crond to resynchronize
and not issue missed commands.  No attempt will be made to issue commands
lost due to a reboot, and commands are not reissued if the previously
issued command is still running.  For example, if you have a crontab
command 'sleep 70' that you wish to run once a minute, cron will only be 
able to issue the command once every two minutes.  If you do not like
this feature, you can run your commands in the background with an '&'.
.PP
The crontab format is roughly similar to that used by vixiecron, but
without complex features.  Individual fields may contain a time, a
time range, a time range with a skip factor, a symbolic range for the
day of week and month in year, and additional subranges delimited with
commas.  Blank lines in the crontab or lines that begin with a hash
(#) are ignored.  If you specify either a month or a day in the month,
and a day of week, the result is effectively ORd... the crontab entry
will be run on the specified day of week and on the specified day in
the month.

.sp 1
.br
# MIN HOUR DAY MONTH DAYOFWEEK   COMMAND
.br
# at 6:10 a.m. every day
.br
10 6 * * * date

.br
# every two hours at the top of the hour
.br
0 */2 * * * date

.br
# every two hours from 11p.m. to 7a.m., and at 8a.m.
.br
0 23-7/2,8 * * * date

.br
# at 11:00 a.m. on the 4th and on every mon, tue, wed
.br
0 11 4 * mon-wed date

.br
# 4:00 a.m. on january 1st
.br
0 4 1 jan * date

.br
# once an hour, every day of january and every Sunday,
all output appended to log file
.br
0 * * jan sun date >>/var/log/messages 2>&1

The command portion of the line is run with 
.B /bin/sh \-c <command>
and may therefore contain any valid bourne shell command.  A common
practice is to run your command with
.B exec
to keep the process table uncluttered.  It is also common to redirect
output to a log file.  If you do not, and the command generates output
on stdout or stderr, the result will be mailed to the user in question.
If you use this mechanism for special users, such as UUCP, you may want
to create an alias for the user to direct the mail to someone else, such
as root or postmaster.
.PP
Internally, this cron uses a quick indexing system to reduce CPU overhead
when looking for commands to execute.  Several hundred crontabs with
several thousand entries can be handled without using noticable CPU 
resources.
.SH BUGS
This implementation of cron does not support the passing of input to
cron jobs on stdin by use of the \fB%\fR sign as specified by the POSIX
standard.
.SH AUTHORS
Matthew Dillon (dillon@apollo.backplane.com), original author.
.br
Gary Langshaw (gary.langshaw@gmail.com), modifications post version 3.2


CHANGELOG
git 5.2 1-July-2026

  * cl_Days[] now a zero based array.

  * main.c renamed crond.c

  * protos.h generation now a little more robust

  * crontab now uses 'replacement' method to write edited files.
    This should prevent issues with editors that replace the file
    when saving.

  * simplify and cleanup RunMainLoop:
    * CheckUpdates() no longer called when a resync is necessary.
    * Remove unnecessary short and long sleeps (stime). sleep
      period is now a constant (WAKEUP_INTERVAL) subject to
      early wakeup owing to SIGNALS.

v5.1 12-June-2026

  * Restore missing month check when running jobs. It had been
    removed by a 3rd party commit taken from Jim's master branch.

  * Crontab resyncs will now happen on the hour instead of every 60
    minutes from the time crond was started.

  * 'cron.update' processing now uses a file lock to avoid race
    conditions.

  * defaults for CRONTABS, CRONSTAMPS and CRONMAIL directory locations
    have been changed as follows
       CRONTABS = /var/lib/cron/crontabs
       CRONSTAMPS = /var/lib/cron/cronstamps
       CRONMAIL = /var/spool/cron

    These differ from their historical locations (but are arguably
    more correct and FHS compliant).

    To continue using the historical locations you will need to
    specify them when running 'make', as follows:
       CRONTABS = /var/spool/cron/crontabs
       CRONSTAMPS = /var/spool/cron/cronstamps
       CRONMAIL = /var/spool/cron/mail

  * other code cleanups and minor fixes. see the git log for
    full details.

v5.0 26-May-2026
  * crond will now clear the process environment before running jobs
    when the '-C' option is specified on startup.  This will avoid the
    need to start the daeon via'env -i' in order to get a pristine
    environment.

  * job output will now be stored in a fixed location, configured via
    the CRONMAIL build option, default: /var/spool/cron/mail rather
    than using a dynamically created directory under TMPDIR.  This
    will prevent repeated daemon stops/starts littering /tmp with
    directories.

  * a new sample script, cronmail-limbo, is provided in the extras/
    directory to handle any mailfiles left behind should the daemon be
    stopped while there are running jobs.

  * crontab now includes signal handlers and a locking mechanism to more
    gracefully handle attempts at concurrent updates to the same crontab.

  * Many other fixes and cleanups. See the git log for details.

v4.5  1-May-2011
  * Some cron jobs were running multiple times. Now we make sure not to
    ArmJobs that are already running; and not to resynchronize while jobs are
    running; and to poll the DST setting. (Fixes Arch FS#18681; thanks to Vincent
    Cappe and Paul Gideon Dann for identifying the second issue; and Tilman
    Sauerbeck for identifying the third.)

  * @monthly was wrongly being parsed the same as @yearly (fixes Arch
    FS#19123). Thanks to Peter Johnson, Paul Gideon Dann, and Tilman Sauerbeck.

  * extra/crond.rc: now uses $CROND_ARGS from /etc/conf.d/crond; sample included
    as extra/crond.conf. Suggested by Eric Bélanger.

  * Running `/etc/rc.d/crond start` after startup could leak unwanted
    environment into cronjobs; now we force crond to start in empty env
    (fixes Arch FS#22085). Thanks to Mantas.

  * Also set LOGNAME environment variable in cronjobs. Requested by Michael
    Trunner; fixes Arch FS#18338.

  * extra/crond.logrotate now correctly gets pid from /var/run/crond.pid
    (fixes Arch FS#18039). Thanks to Kay Abendroth, revel, and Chlump Chatkupt.

  * extra/prune-cronstamps now only deletes files, and is formatted as a
    @weekly crontab. Thanks to Alec Moskvin <alecm@gmx.com>.

  * extra/crontab.vim works around an issue where vim's writebackup would
    interfere with crontab's security model (addresses Arch FS#18352).
    Thanks to Armadillo and Simon Williams.

  * Makefile uses $LDFLAGS (fixes Arch FS#23784). Thanks to Kristoffer Tidemann
    and Mike Frysinger.

  * defs.h sets default locations for CRONTABS and CRONSTAMPS beneath /var/spool/cron/,
    as in earlier versions of dcron.

  * Documentation updates.

  * Thanks for testing and feedback: Feifei Jia, Spider.007, Ray Kohler,
    Igor Zakharoff, Edward Hades, and Joe Lightning.

v4.4  17-Jan-2010
  * Finished mailjobs were being left as zombie processes. Fixed.

  * When using crond with logging-to-file, user jobs could only log some
    events if they had write access to the log. Fixed this by having crond
    keep a file descriptor open to the log; also added a SIGHUP handler
    to make crond re-open the logfile. The sample logrotate script now
    sends that signal.

  * More sensible command-line parsing by crontab.

  * Add prune-cronstamps to extra; document extra/*; general improvement
    of README and manpages.

  * Portability improvements, and defs.h now has fuller comments about
    requirements.

  * Makefile improvements: `make` now caches variables for `make install`;
    don't stomp CFLAGS environment variable, and added BINDIR,SBINDIR,MANDIR.

  * Thanks to Juergen Daubert for more testing and suggestions.

v4.3  11-Jan-2010
  * Internal refactoring to make buffer overflow checks
    clearer and portability issues more explicit.

  * Made file argument to -L mandatory; optional args to
    getopt needs GNU extensions.

  * Makefile tweaks. Added CRONTAB_GROUP for `make install`.
    Renamed TIMESTAMPS -> CRONSTAMPS.

  * Thanks to Juergen Daubert for testing and suggestions.

v4.2  11-Jan-2010
  * Makefile tweaks; moved more constants to #defines.

v4.1  10-Jan-2010
  * Fixed bug in parsing some numeric fields in crontabs. (Terminus of range
    wasn't being modded.)

  * Updated Makefile to make it easier to customize timestamps at configure
    time. Also, if LC_TIME is defined when crond runs, we use that instead of
    compiled-in default (for logging to files, to customize syslog output use
    syslog-ng's 'template' command).


crontab.c
//  crontab.c
//
//  Copyright 1994 Matthew Dillon (dillon@apollo.backplane.com)
//  Copyright 2026 Gary Langshaw (gary.langshaw@gmail.com)
//  May be distributed under the GNU General Public License

#include "chuser.h"
#include "defs.h"
#include "subs.h"

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <syslog.h>
#include <unistd.h>

const char *CDir = CRONTABS;
const char *CDirOpt = NULL;
int UserId;
int LogLevel = 9;

void EditFile(const char *user, const char *file);
int GetReplaceStream(const char *user, const char *file);

int main(int argc, char *argv[])
{
  enum
  {
    NONE,
    EDIT,
    LIST,
    REPLACE,
    DELETE
  } option = NONE;
  struct passwd *pas;
  char edFile[] = TMPDIR "/crontab.XXXXXX";
  char *repFileName = NULL;
  char *userName = NULL;
  int repFd = 0;
  char caller[256]; /* user that ran program */
  char buf[BUFSIZ];
  int opt;

  UserId = getuid();
  if ((pas = getpwuid(UserId)) == NULL)
  {
    perror("getpwuid");
    exit(1);
  }
  snprintf(caller, sizeof(caller), "%s", pas->pw_name);

#define ERRMSG_EXCLUSIVE_OPTION "only one of: -r, -e, -l can be specified."
  while ((opt = getopt(argc, argv, "ldreu:")) != -1)
  {
    if (opt == '?')
      errx(1, "aborted. unrecognised option");
    switch (opt)
    {
    case 'c':
      CDirOpt = optarg;
      break;
    case 'u':
      userName = optarg;
      break;
    case 'e':
      if (option != NONE)
        errx(1, ERRMSG_EXCLUSIVE_OPTION);
      option = EDIT;
      break;
    case 'l':
      if (option != NONE)
        errx(1, ERRMSG_EXCLUSIVE_OPTION);
      option = LIST;
      break;
    case 'd': /* for backward compatibility, POSIX option is -r */
    case 'r':
      if (option != NONE)
        errx(1, ERRMSG_EXCLUSIVE_OPTION);
      option = DELETE;
      break;
    default:
      break;
    }
  }
#undef ERRMSG_EXCLUSIVE_OPTION

  if (optind != argc && (option == LIST || option == EDIT || option == DELETE))
    errx(1, "unexpected FILE argument: %s\n", argv[optind]);

  if (argc > optind + 1)
    errx(1, "too many arguments\n");

  if (option == NONE)
  {
    option = REPLACE;
    if (optind == argc - 1)
      repFileName = argv[optind];
    else
      repFileName = "-";
  }

  if (userName)
  {
    if (getuid() == (uid_t)0)
    {
      pas = getpwnam(userName);
      if (pas)
      {
        UserId = pas->pw_uid;
      }
      else
      {
        errx(1, "user %s unknown\n", userName);
      }
    }
    else
      errx(1, "only the superuser may specify a user\n");
  }

  if (CDirOpt)
  {
    if (getuid() == (uid_t)0)
      CDir = CDirOpt;
    else
      errx(1, "-c option: superuser only\n");
  }

  //  Get password entry

  if ((pas = getpwuid(UserId)) == NULL)
  {
    perror("getpwuid");
    exit(1);
  }

  //  Change directory to our crontab directory

  if (chdir(CDir) < 0)
    errx(1, "cannot change dir to %s: %s\n", CDir, strerror(errno));

  //  Handle options as appropriate

  switch (option)
  {
  case LIST:
  {
    FILE *fi;

    if ((fi = fopen(pas->pw_name, "r")))
    {
      while (fgets(buf, sizeof(buf), fi) != NULL)
        fputs(buf, stdout);
      fclose(fi);
    }
    else
    {
      fprintf(stderr, "no crontab for %s\n", pas->pw_name);
    }
  }
  break;
  case EDIT:
  {
    FILE *fi;
    int fd;
    int n;

    if ((fd = mkstemp(edFile)) >= 0)
    {
      chown(edFile, getuid(), getgid());
      if ((fi = fopen(pas->pw_name, "r")))
      {
        while ((n = fread(buf, 1, sizeof(buf), fi)) > 0)
          write(fd, buf, n);
      }
      close(fd);
      EditFile(caller, edFile);
      repFileName = edFile;
    }
    else
    {
      errx(1, "unable to create %s\n", edFile);
    }
  }
    option = REPLACE;
    // fall through.
  case REPLACE:
  {
    char path[PATH_MAX];
    int fd;
    int n;

    //  If there is a replacement file, obtain a secure descriptor to it.
    if (repFileName)
    {
      repFd = GetReplaceStream(caller, repFileName);
      if (repFileName == edFile)
        remove(edFile);
      if (repFd < 0)
      {
        errx(1, "unable to read replacement file\n");
      }
    }

    snprintf(path, sizeof(path), ".%s.XXXXXX", pas->pw_name);
    if ((fd = mkstemp(path)) >= 0)
    {
      while ((n = read(repFd, buf, sizeof(buf))) > 0)
      {
        write(fd, buf, n);
      }
      close(fd);
      rename(path, pas->pw_name);
    }
    else
    {
      fprintf(stderr, "unable to create %s/%s: %s\n", CDir, path,
              strerror(errno));
    }
    close(repFd);
  }
  break;
  case DELETE:
    remove(pas->pw_name);
    break;
  case NONE:
  default:
    break;
  }

  //   Bump notification file.  Handle window where crond picks file
  //   up before we can write our entry out.

  if (option == REPLACE || option == DELETE)
  {
    FILE *fo;
    struct stat st;

    while ((fo = fopen(CRONUPDATE, "a")))
    {
      fprintf(fo, "%s\n", pas->pw_name);
      fflush(fo);
      if (fstat(fileno(fo), &st) != 0 || st.st_nlink != 0)
      {
        fclose(fo);
        break;
      }
      fclose(fo);
    }
    if (fo == NULL)
    {
      fprintf(stderr, "unable to append to %s/%s\n", CDir, CRONUPDATE);
    }
  }
  (volatile void)exit(0);
  // not reached.
}

int GetReplaceStream(const char *user, const char *file)
{
  int filedes[2];
  int pid;
  int fd;
  int n;
  char buf[PIPE_BUF];

  if (pipe(filedes) < 0)
  {
    perror("pipe");
    return (-1);
  }
  if ((pid = fork()) < 0)
  {
    perror("fork");
    return (-1);
  }
  if (pid > 0)
  {
    //  PARENT
    close(filedes[1]);
    if (read(filedes[0], buf, 1) != 1)
    {
      close(filedes[0]);
      filedes[0] = -1;
    }
    return (filedes[0]);
  }

  //  CHILD

  close(filedes[0]);

  if (ChangeUser(user, 0) < 0)
    errx(EXIT_FAILURE, "crontab: failed to change user to %s\n", user);

  if (strcmp("-", file) == 0)
    fd = 0;
  else
    fd = open(file, O_RDONLY);

  if (fd < 0)
    errx(EXIT_FAILURE, "unable to open %s\n", file);

  buf[0] = 0;
  write(filedes[1], buf, 1);
  while ((n = read(fd, buf, sizeof(buf))) > 0)
  {
    write(filedes[1], buf, n);
  }
  exit(0);
}

void EditFile(const char *user, const char *file)
{
  int pid;

  if ((pid = fork()) == 0)
  {
    //  CHILD - change user and run editor
    const char *ptr;
    char visual[1024];

    if (ChangeUser(user, 1) < 0)
      errx(EXIT_FAILURE, "crontab: failed to change user to %s\n", user);

    if ((ptr = getenv("VISUAL")) == NULL || strlen(ptr) > 256)
      ptr = PATH_VI;

    snprintf(visual, sizeof(visual), "%s %s", ptr, file);
    execl("/bin/sh", "/bin/sh", "-c", visual, NULL);
    perror("exec");
    exit(0);
  }
  if (pid < 0)
  {
    //  PARENT - failure
    perror("fork");
    exit(1);
  }
  waitpid(pid, NULL, 0);
}
